### DISCUSSION TITLE
config: Creating missing directories in `CONFIG_FILE` path

### DISCUSSION THREAD

**Discussion by dabrady** [bashly-framework/bashly#377] [Ideas]
Aloha 🌸 

I'm enhancing my CLI to leverage a 'dot config directory', similar to many other CLIs. The standard `config` library will currently fail to create the `CONFIG_FILE` if any of part of the `CONFIG_FILE` path is missing, not just if the file is missing. Here is a minimal example:
```shell
example-cli
➜  bashly init
created src/bashly.yml
run bashly generate to create the bash script

example-cli
➜  bashly add config
created src/lib/config.sh

example-cli
➜  echo 'config_show' > src/download_command.sh

example-cli
➜  bashly generate
creating user files in src
created src/initialize.sh
skipped src/download_command.sh (exists)
created src/upload_command.sh
created ./cli
run ./cli --help to test your bash script

example-cli
➜  CONFIG_FILE='.example/config.ini' ./cli download foo
touch: .example/config.ini: No such file or directory
```

As a patch for my use-case, I have modified the `config_init` function in `src/lib/config.sh` such that it calls `mkdir -p $(dirname $CONFIG_FILE)` prior to attempting to `touch` the `CONFIG_FILE`:
```diff
diff --git a/src/lib/config.sh b/src/lib/config.sh
index bb3074d..eac2487 100644
--- a/src/lib/config.sh
+++ b/src/lib/config.sh
@@ -13,7 +13,7 @@
 ##
 config_init() {
   CONFIG_FILE=${CONFIG_FILE:=config.ini}
-  [[ -f "$CONFIG_FILE" ]] || touch "$CONFIG_FILE"
+  [[ -f "$CONFIG_FILE" ]] || ( mkdir -p "$(dirname $CONFIG_FILE)" && touch "$CONFIG_FILE" )
 }
```

Do you think this is a valuable enhancement to the core library, or are there good reasons to avoid that?

**Comment by DannyBen** [comment]
A few notes:

1. Most of the libraries added with `bashly add ...` are intended to be modified if needed. This is why they are added to user-controlled space instead of just being injected to the resulting script.
2. If you happen to edit a library file, be sure to delete the magic comment on top: `[@bashly-upgrade <library_name>]`. With this comment, when you run `bashly generate --upgrade`, this file will be forcibly overwritten by the factory default.
3. Another place that you can ensure the config path exists, without modifying the library, is the `src/initialize.sh` hook (`bashly add hooks`)
4. I intentionally avoided `mkdir -p`, to avoid surprises that will force the user to manually delete something that was accidentally created by the script.
